src/iso9660.rs

```diff
@@ -0,0 +1,364 @@
+        Ok(primary.root.clone())
+    }
+
+    /// Returns an iterator over the records of a directory.
+    pub fn list_dir(&mut self, dir: &Directory) -> Result<IsoFsIterator> {
+        seek_to_sector(&mut self.file, dir.address)
+            .with_context(|| format!("seeking to directory {}", dir.name))?;
+
+        let mut buf = vec![0; dir.length as usize];
+        self.file
+            .read_exact(&mut buf)
+            .with_context(|| format!("reading directory {}", dir.name))?;
+
+        Ok(IsoFsIterator {
+            dir: Bytes::from(buf),
+        })
+    }
+
+    /// Returns the record for a specific path.
+    pub fn get_path(&mut self, path: &str) -> Result<DirectoryRecord> {
+        let as_path = Path::new(&path);
+        let mut parent_dir = as_path
+            .parent()
+            .ok_or_else(|| anyhow!("path {} has no parent", &path))?;
+        if parent_dir.has_root() {
+            parent_dir = parent_dir
+                .strip_prefix("/")
+                .with_context(|| format!("making path '{}' relative", path))?;
+        }
+        let filename = as_path
+            .file_name()
+            .ok_or_else(|| anyhow!("path {} has no base", &path))?;
+        let filename = filename.to_str().unwrap(); // `path` is &str
+
+        let mut dir = self.get_root_directory()?;
+        for component in parent_dir.components() {
+            if let std::path::Component::Normal(c) = component {
+                let c = c.to_str().unwrap(); // `path` is &str
+                match self.get_dir_record(&dir, c)? {
+                    DirectoryRecord::Directory(d) => dir = d,
+                    DirectoryRecord::File(_) => {
+                        bail!("component {:?} in path {} is not a directory", c, &path)
+                    }
+                }
+            } else {
+                bail!("path is not canonical: {}", &path);
+            }
+        }
+
+        self.get_dir_record(&dir, filename)
+    }
+
+    /// Returns the record for a specific name in a directory.
+    pub fn get_dir_record(&mut self, dir: &Directory, name: &str) -> Result<DirectoryRecord> {
+        for record in self
+            .list_dir(dir)
+            .with_context(|| format!("listing directory {}", dir.name))?
+        {
+            let record = record?;
+            match &record {
+                DirectoryRecord::Directory(d) if d.name == name => return Ok(record),
+                DirectoryRecord::File(f) if f.name == name => return Ok(record),
+                _ => continue,
+            }
+        }
+        Err(anyhow!("directory {} has no record for {}", dir.name, name))
+    }
+
+    /// Returns a reader for a file record.
+    pub fn read_file(&mut self, file: &File) -> Result<impl Read + '_> {
+        seek_to_sector(&mut self.file, file.address)
+            .with_context(|| format!("seeking to file {}", file.name))?;
+        Ok(LimitReader::new(&self.file, file.length as u64, None))
```

Comment:
I just noticed that `Read.take()` exists, so we could use it and revert da13fe03ad53aab1838f31dbd9a6ab34757039bf.

Reply:
Also, should we wrap this in a BufReader?

Reply:
Let's tackle the `Read.take()` thing separately.

Added `BufReader` wrapper.

Reply:
Turns out that `take()` consumes the `File` and `File.try_clone()` returns a clone that shares the file offset with the original, so I'm thinking the `take()` change isn't worth it.

Reply:
I thought this too at first, but it can work if we wrap the `LimitReader` itself, but it's kinda awkward since they essentially do almost the same thing.

Reply:
Ahh, yeah, good call.